### PR TITLE
compiler: add simd_ctpop intrinsic

### PR DESCRIPTION
**Pull request by workingjubilee**
Fairly straightforward addition.

cc @rust-lang/opsem new (extremely boring) intrinsic

### CODE REVIEW

**Comment by workingjubilee** [comment]
Yes, that's the idea.

File: compiler/rustc_codegen_llvm/src/intrinsic.rs [hunk 1]
```diff
@@ -2354,31 +2357,38 @@ fn generic_simd_intrinsic<'ll, 'tcx>(
...
+                // this fun bonus i1 arg means "poison if the arg vector contains zero"
+                let fn_ty = bx.type_func(&[vec_ty, bx.type_i1()], vec_ty);
+                let f = bx.declare_cfn(llvm_intrinsic, llvm::UnnamedAddr::No, fn_ty);
+                Ok(bx.call(
+                    fn_ty,
+                    None,
+                    None,
+                    f,
+                    // simd_ctlz and simd_cttz are exposed to safe code, so let's not poison anything
+                    &[args[0].immediate(), bx.const_int(bx.type_i1(), 0)],
```
Comment (workingjubilee) [inline compiler/rustc_codegen_llvm/src/intrinsic.rs:2380]: I found the existing comments insufficient to explain why `ctlz` and `cttz` had such a peculiar handling. "You know, if we pass a different value for this bool, it makes the resulting program UB" _is_ fairly common in the far backend, but that doesn't mean it should go unremarked.

**Review by calebzulawski — APPROVED** [review]

**Comment by calebzulawski** [comment]
r? @RalfJung

looks good to me

**Comment by saethlin** [comment]
I presume it's implausible for this intrinsic to have a fallback body because we can't actually write the generic bound?

Can we then add Miri support? I feel like it shouldn't be too hard, the boilerplate is already written out: https://github.com/rust-lang/rust/blob/bfa3635df920454cdf03f6d268dfaac769375df3/src/tools/miri/src/intrinsics/simd.rs#L18-L20

**Comment by workingjubilee** [comment]
> I presume it's implausible for this intrinsic to have a fallback body because we can't actually write the generic bound?
> 
> Can we then add Miri support? I feel like it shouldn't be too hard, the boilerplate is already reusable in this function:
> 
> https://github.com/rust-lang/rust/blob/bfa3635df920454cdf03f6d268dfaac769375df3/src/tools/miri/src/intrinsics/simd.rs#L18-L20

I can add Miri support yeah.

I'm not sure what adding a fallback body would look like here, it might be possible to twist the code such that it actually works, tbh.

**Comment by workingjubilee** [comment]
@saethlin hmm so wait I looked around and didn't see a magic wand to wave... am I supposed to just, for a fallback body, literally write a body for the intrinsic?

lemme think about that, it might work.

**Comment by workingjubilee** [comment]
hm. I know how to add the appropriate generic bounds, they just might make someone mad.

File: compiler/rustc_codegen_llvm/src/intrinsic.rs (hunk 1 above)
Comment (RalfJung) [reply compiler/rustc_codegen_llvm/src/intrinsic.rs:2380]: It's not clear to me which part of the next line that "not poison anything" refers to, though. I guess it's the "0"? What does it mean?

**Comment by RalfJung** [comment]
Happy to review a Miri implementation, but LLVM codegen is not something I can confidently review.
r? @nikic

We haven't yet used a fallback body for any SIMD intrinsic so maybe this is not the right time and place for that experiment. :) (It could delay the PR quite a bit, trying to get that to work.)

**Comment by bjorn3** [comment]
Would you mind adding cg_clif support too at https://github.com/rust-lang/rust/blob/7c735958b2c5c25565625820134d82b0871a426e/compiler/rustc_codegen_cranelift/src/intrinsics/simd.rs#L347? Cranelift calls the instruction that needs to be used here popcnt: https://github.com/rust-lang/rust/blob/7c735958b2c5c25565625820134d82b0871a426e/compiler/rustc_codegen_cranelift/src/intrinsics/mod.rs#L649

**Comment by calebzulawski** [comment]
> hm. I know how to add the appropriate generic bounds, they just might make someone mad.

They will leak into the `std::simd` interface, which we don't want

**Comment by workingjubilee** [comment]
> > hm. I know how to add the appropriate generic bounds, they just might make someone mad.
> 
> They will leak into the `std::simd` interface, which we don't want

Ehn, we already have the relevant bounds or approximately so.

As Ralf said, it would delay things significantly.

Yeah, I noticed that `cg_clif` and `cg_gcc` were already parsing the LLVM intrinsic, which made me surprised no one added this yet. :eyes:

I'll add the followups.

**Comment by RalfJung** [comment]
Miri support should be pretty easy -- just another case [here](https://github.com/rust-lang/rust/blob/5e41ff516fad8f2984cc0c290948c458970e4a86/src/tools/miri/src/intrinsics/simd.rs#L70-L73) that maps to `Op::Numeric(sym::ctpop)`. This completes the support for the "numeric" intrinsics, except for the nonzero variants.

For tests I think [this](https://github.com/rust-lang/rust/blob/3046dbe7deaebf2804fc822dd1828b3c794edc6f/src/tools/miri/tests/pass/intrinsics/portable-simd.rs#L491) function would be good.

**Comment by workingjubilee** [comment]
I briefly glanced at cg_gcc but I do not understand the relevant code for the popcount implementation, which appears to have many edge-cases, and does not seem to dispatch to the fast vpopcount builtins anyways. I hope it won't be too hard to add, but my apologies.

File: compiler/rustc_codegen_llvm/src/intrinsic.rs (hunk 1 above)
Comment (workingjubilee) [reply compiler/rustc_codegen_llvm/src/intrinsic.rs:2380]: @RalfJung Yes, let me clarify that...

**Comment by saethlin** [comment]
Neat, thanks for the Miri implementation.

**Comment by RalfJung** [comment]
r=me on the Miri implementation and test, thanks. :)

File: compiler/rustc_codegen_llvm/src/intrinsic.rs [hunk 2]
```diff
@@ -2368,16 +2368,16 @@ fn generic_simd_intrinsic<'ll, 'tcx>(
             // byte swap is no-op for i8/u8
             sym::simd_bswap if int_size == 8 => Ok(args[0].immediate()),
             sym::simd_ctlz | sym::simd_cttz => {
-                // this fun bonus i1 arg means "poison if the arg vector contains zero"
+                // for the (int, i1 immediate) pair, the second arg adds `(0, true) => poison`
                 let fn_ty = bx.type_func(&[vec_ty, bx.type_i1()], vec_ty);
+                let dont_poison_on_zero = bx.const_int(bx.type_i1(), 0);
```
Comment (RalfJung) [inline compiler/rustc_codegen_llvm/src/intrinsic.rs:2373]: Shouldn't this be called `poison_on_zero`, which we are then setting to `false`?

Comment (workingjubilee) [reply compiler/rustc_codegen_llvm/src/intrinsic.rs:2373]: hmmm, plausible. I was going for a "nah nah" instead of a "yeah nah".

**Pull request merged**